src/openms/source/METADATA/ID/IdentificationDataConverter.cpp

```diff
@@ -0,0 +1,888 @@
+      }
+      IdentificationData::DataQueryRef query_ref =
+        id_data.registerDataQuery(query);
+
+      IdentificationData::ScoreType score_type(
+        pep.getScoreType(), pep.isHigherScoreBetter(), step_ref->software_ref);
+      IdentificationData::ScoreTypeRef score_ref =
+        id_data.registerScoreType(score_type);
+
+      // PeptideHit:
+      for (const PeptideHit& hit : pep.getHits())
+      {
+        if (hit.getSequence().empty()) continue;
+        IdentificationData::IdentifiedPeptide peptide(hit.getSequence());
+        peptide.processing_step_refs.push_back(step_ref);
+        for (const PeptideEvidence& evidence : hit.getPeptideEvidences())
+        {
+          const String& accession = evidence.getProteinAccession();
+          if (accession.empty()) continue;
+          IdentificationData::ParentMolecule parent(accession);
+          parent.processing_step_refs.push_back(step_ref);
+          // this will merge information if the protein already exists:
+          IdentificationData::ParentMoleculeRef parent_ref =
+            id_data.registerParentMolecule(parent);
+          IdentificationData::MoleculeParentMatch match(
+            evidence.getStart(), evidence.getEnd(), evidence.getAABefore(),
+            evidence.getAAAfter());
+          peptide.parent_matches[parent_ref].insert(match);
+        }
+        IdentificationData::IdentifiedPeptideRef peptide_ref =
+          id_data.registerIdentifiedPeptide(peptide);
+
+        IdentificationData::MoleculeQueryMatch match(peptide_ref, query_ref);
+        match.charge = hit.getCharge();
+        static_cast<MetaInfoInterface&>(match) = hit;
+        if (!hit.getPeakAnnotations().empty())
+        {
+          match.peak_annotations[step_ref] = hit.getPeakAnnotations();
+        }
+        match.processing_step_refs.push_back(step_ref);
+
+        // analysis results from pepXML:
+        for (const PeptideHit::PepXMLAnalysisResult& ana_res :
+               hit.getAnalysisResults())
+        {
+          Software software;
+          software.setName(ana_res.score_type); // e.g. "peptideprophet"
+          IdentificationData::ProcessingSoftwareRef software_ref =
+            id_data.registerDataProcessingSoftware(software);
+          IdentificationData::DataProcessingStep sub_step(software_ref);
+          if (query.input_file_opt)
+          {
+            sub_step.input_file_refs.push_back(*query.input_file_opt);
+          }
+          IdentificationData::ProcessingStepRef sub_step_ref =
+            id_data.registerDataProcessingStep(sub_step);
+          match.processing_step_refs.push_back(sub_step_ref);
+          for (const pair<String, double>& sub_pair : ana_res.sub_scores)
+          {
+            IdentificationData::ScoreType sub_score;
+            sub_score.name = sub_pair.first;
+            sub_score.software_opt = software_ref;
+            IdentificationData::ScoreTypeRef sub_score_ref =
+              id_data.registerScoreType(sub_score);
```

Comment:
the same for the scores, you keep a global list of the names of the scores and then you only store a `pair` of `<ref, double` to reference the score? I assume that is way more efficient than using the MetaValue ?

Reply:
Is that is the case, I am wondering why not simply store a `std::pair<int, double>` where the `int` references the entry in a global list of scores. I believe the interface may be a bit cleaner, even though you introduce another set of indirection but there is less dependency on the actual type of `IdentificationData::ScoreTypeRef` and all you would need is a function `IdentificationData::getScore(int score_id)` to retrieve the score, its name etc. Also it may be even more memory efficient as an `int` is only 4 bytes whereas an iterator is usually the same size as a ptr = 8 bytes

Reply:
I haven't compared against `MetaValue` specifically.

> why not simply store a std::pair<int, double>

As you said, mainly due to the additional indirection. I also think that using `ScoreTypeRef` makes things more explicit than having `int`-that-indexes-a-list-of-scores.